### ISSUE TITLE
Double Payment investigation

### ISSUE THREAD

**Issue by GriffGreen** [Giveth/giveth-dapp#1789]
Looking at the bridge it seems like this payment was made twice:
https://beta.giveth.io/campaigns/5d110631994eac53adc8f21e/milestones/5f510a4b0573fc732be1c67f
1 milestone payment request:
<img width="638" alt="Screen Shot 2021-02-11 at 7 14 16 AM" src="https://user-images.githubusercontent.com/14945613/107635551-9ca08c80-6c6b-11eb-9321-150b75309345.png">

2 milestone payouts:

<img width="694" alt="Screen Shot 2021-02-11 at 7 20 00 AM" src="https://user-images.githubusercontent.com/14945613/107635708-e38e8200-6c6b-11eb-9126-1b7c1e833645.png">

<img width="701" alt="Screen Shot 2021-02-11 at 7 19 50 AM" src="https://user-images.githubusercontent.com/14945613/107635713-e4bfaf00-6c6b-11eb-99a5-5764552b8e6f.png">

Luckily this was to Kris and I am sure he made money off the DAI sitting in the bridge.... he will give it back.

The more important piece is understanding how and why it happened so it doesn't happen again.... did we make a manual tx?

**Comment by mohammadranjbarz** [comment]
@aminlatifi 
I ran this query on database:

```

db.getCollection('homepaymentstransactions').aggregate(
  { $group: {
      // Group by fields to match on (a,b)
      _id: { traceId: "$traceId",
        donationTxHash: "$donationTxHash" ,
        event: "$event",
        "tokens": "$payments.symbol",
      },

      // Count number of matching docs for the group
      count: { $sum:  1 },

      // Save the _id for matching docs
      docs: { $push: {
          payments:"$$ROOT.payments",
          timestamp: "$$ROOT.timestamp"
          }}
    }},

  // Limit results to duplicates (more than 1 match)
  { $match: {
      count: { $gt : 1 }
    }}
)
```

and this is the result:
```
/* 1 */
{
    "_id" : {
        "traceId" : "5e9c60b3f597dd6cca6436fe",
        "donationTxHash" : "0xabde2f9dee77d149e8c8365daa3f4e7207e1d28af396eacee13cbedf0d8d57ff",
        "event" : "PaymentExecuted",
        "tokens" : [ 
            "DAI"
        ]
    },
    "count" : 2.0,
    "docs" : [ 
        {
            "payments" : [ 
                {
                    "_id" : ObjectId("60c1f5589e25f26ba4585cf3"),
                    "amount" : "605970000000000005968",
                    "symbol" : "DAI"
                }
            ],
            "timestamp" : ISODate("2020-05-10T08:59:41.000Z")
        }, 
        {
            "payments" : [ 
                {
                    "_id" : ObjectId("60d78054cfcae50bb5852311"),
                    "amount" : "605970000000000005968",
                    "symbol" : "DAI"
                }
            ],
            "timestamp" : ISODate("2021-06-26T19:29:35.000Z")
        }
    ]
}

/* 2 */
{
    "_id" : {
        "traceId" : "5f510a4b0573fc732be1c67f",
        "donationTxHash" : "0x05da9b3441e04331f5a829dc1001b69665e5794f23b0a9a15a972ad10e2183f4",
        "event" : "PaymentAuthorized",
        "tokens" : [ 
            "DAI"
        ]
    },
    "count" : 2.0,
    "docs" : [ 
        {
            "payments" : [ 
                {
                    "_id" : ObjectId("60c1f5b19e25f26ba4585ef6"),
                    "amount" : "3565000000000000000000",
                    "symbol" : "DAI"
                }
            ],
            "timestamp" : ISODate("2020-09-08T11:49:40.000Z")
        }, 
        {
            "payments" : [ 
                {
                    "_id" : ObjectId("60c1f5b29e25f26ba4585f00"),
                    "amount" : "3565000000000000000000",
                    "symbol" : "DAI"
                }
            ],
            "timestamp" : ISODate("2020-09-14T14:16:50.000Z")
        }
    ]
}

/* 3 */
{
    "_id" : {
        "traceId" : "5e9c60b3f597dd6cca6436fe",
        "donationTxHash" : "0xabde2f9dee77d149e8c8365daa3f4e7207e1d28af396eacee13cbedf0d8d57ff",
        "event" : "PaymentAuthorized",
        "tokens" : [ 
            "DAI"
        ]
    },
    "count" : 2.0,
    "docs" : [ 
        {
            "payments" : [ 
                {
                    "_id" : ObjectId("60c1f5579e25f26ba4585ced"),
                    "amount" : "605970000000000005968",
                    "symbol" : "DAI"
                }
            ],
            "timestamp" : ISODate("2020-05-08T16:29:33.000Z")
        }, 
        {
            "payments" : [ 
                {
                    "_id" : ObjectId("60c1f70b9e25f26ba4586632"),
                    "amount" : "605970000000000005968",
                    "symbol" : "DAI"
                }
            ],
            "timestamp" : ISODate("2021-05-26T12:04:00.000Z")
        }
    ]
}
```
This has two donations with same amount, and it was ok 
https://trace.giveth.io/trace/nelson-smart-contract-work-on-cstk-token-dapp

and this is the Kris trace that had double payment
https://trace.giveth.io/trace/kris-august-20-work

so we just had one double payment

**Comment by aminlatifi** [comment]
Kris Trace double payment is canceled

![image](https://user-images.githubusercontent.com/5684607/123912320-238fa080-d992-11eb-9747-a7aa1cc7b107.png)

**Comment by aminlatifi** [comment]
@GriffGreen Above shows double events for single payment. Kris payment doesn't have double `PaymentExecuted` event and it means bridge had created its job two times, but finally it didn't end up double transfer to Kris wallet.

But Nelson had received double payments for single collect of 605.97 DAI ("paymentId" : "968" and "paymentId" : "1361")
https://trace.giveth.io/trace/nelson-smart-contract-work-on-cstk-token-dapp

**Comment by aminlatifi** [comment]
Thanks @mohammadranjbarz for the query

**Comment by GriffGreen** [comment]
hmmm so there was a bug here and he was able to collect 605.97 DAI more than the milestone received in donations?

@aminlatifi @mohammadranjbarz can you confirm i got that right?

If so, Great Catch!!!! How did the bridge not catch it? How can we make sure the bridge does catch it?

**Comment by GriffGreen** [comment]
This was 3 days ago!?

https://etherscan.io/tx/0xa83830df727ddf5e05c8a66ea31df9e9b8fd1d214512a323f0e0600ace8f3d43

**Comment by GriffGreen** [comment]
oh ok! It did catch it

<img width="738" alt="Screen Shot 2021-06-30 at 8 55 45 AM" src="https://user-images.githubusercontent.com/14945613/123915673-0900fb80-d981-11eb-9555-cf81deab666d.png">

Looks like it slipped by @bowensanders

**Comment by GriffGreen** [comment]
I will reach out to nelson and see if he tried to collect again

**Comment by mohammadranjbarz** [comment]
> I will reach out to nelson and see if he tried to collect again

@GriffGreen 
Nelson's trace didn't have problem, he has two exact amount collect in single trace so his payments came out from query result, 
But I think it didn't have bug ( he collected twice in different times, so he get money twice and it's ok)

 https://trace.giveth.io/trace/nelson-smart-contract-work-on-cstk-token-dapp
![Screen Shot 1400-04-09 at 11 30 12](https://user-images.githubusercontent.com/9850545/123916279-b7636b80-d996-11eb-8399-df3494c33a32.png)

**Comment by GriffGreen** [comment]
it looks like to me that he collected

605.97+1194.09+605.97 = 2,406.03 DAI

and the milestone says he only had 2220.06 DAI donated...

So one of those numbers is wrong

**Comment by GriffGreen** [comment]
@mohammadranjbarz ^^

**Comment by aminlatifi** [comment]
> hmmm so there was a bug here and he was able to collect 605.97 DAI more than the milestone received in donations?
> 
> @aminlatifi @mohammadranjbarz can you confirm i got that right?
> 
> If so, Great Catch!!!! How did the bridge not catch it? How can we make sure the bridge does catch it?

![Uploading image.png…]()

**Comment by aminlatifi** [comment]
> I will reach out to nelson and see if he tried to collect again

It's not about his try to collect multiple time, bridge had created double tx for his payment

**Comment by aminlatifi** [comment]
@GriffGreen That's the one I did a mistake and told you and Bowen about it

**Comment by aminlatifi** [comment]
![image](https://user-images.githubusercontent.com/5684607/123918583-43769280-d999-11eb-913a-4cf91be7fbef.png)

**Comment by mohammadranjbarz** [comment]
> it looks like to me that he collected
> 
> 605.97+1194.09+605.97 = 2,406.03 DAI
> 
> and the milestone says he only had 2220.06 DAI donated...
> 
> So one of those numbers is wrong
@GriffGreen 
You are right,  my mistake, Nelson had one collect for 605 and two bridge payment, that Amin said the cause in the above message

**Comment by bowensanders** [comment]
This seems to be the one that i DID delay a month when I was asked to, but I did not check back on it because I had assumed, once you asked me to delay it by 30 days, that the payment cancellation via multisig would have gone through, even though I wasn't asked to participate in the cancellation :(

**Comment by bowensanders** [comment]
to further iterate I did actually delay that payment. 
<img width="1680" alt="Screen Shot 2021-05-28 at 9 14 08 PM" src="https://user-images.githubusercontent.com/27023908/124217894-05b75e00-daae-11eb-90ca-a64a02e2cb69.png">
Glad I kept the screenshot.
Sorry I didn't catch it the second time around, *that* part is on me, though I figured I would have been notified that appropriate action hadn't been taken on the back end of this, and for that I apologize.
